src/heif.imageio/heifinput.cpp

```diff
@@ -402,7 +449,19 @@ HeifInput::read_native_scanline(int subimage, int miplevel, int y, int /*z*/,
         return false;
     }
     hdata += (y - m_spec.y) * ystride;
-    memcpy(data, hdata, m_spec.width * m_spec.pixel_bytes());
+    if (m_spec.format == TypeUInt16) {
+        // Convert from 10 or 12 bits to 16 bits. Read little endian
+        // zero padded bits, and shift to scale up.
+        const uint8_t* in       = hdata;
+        uint16_t* out           = static_cast<uint16_t*>(data);
+        const int bitshift      = 16 - m_bitdepth;
+        const size_t num_values = m_spec.width * m_spec.nchannels;
+        for (size_t i = 0; i < num_values; i++, out++, in += 2) {
+            *out = uint16_t((in[1] << 8) | in[0]) << bitshift;
+        }
```

Comment:
Are you sure this is correct? If you're converting merely by shifting, that means that for 10 bits (for example), a fully saturated 0b1111111111 will turn into 16 bit 0b1111111111000000, but really you want it to be 0b1111111111111111, so that it's truly 1.0.

We have a utility in fmath.h called bit_range_convert that does this correctly, and you can see it used in several of the other readers like targa, rla, jpeg2000, etc.

Reply:
Switched to using `bit_range_convert` now.